File: Python/Classification/Logistic_Regression.py
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

dataset = pd.read_csv('dataset.csv')
x = dataset.iloc[:, [2,3]].values # Characteristics vector
y = dataset.iloc[:, 4].values # Predictable vector

from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state = 0)

""" If necesary
from sklearn.preprocessing import StandardScaler
sc_x = StandardScaler()
x_train = sc_x.fit_transform(x_train)
x_test = sc_x.transform(x_test)
"""

# Adjust logistic regression model on the training set
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression(random_state = 0)
classifier.fit(x_train, y_train)

# Predict results with testing set
y_pred = classifier.predict(x_test)

# Create a Confusion Matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)

"""
 cm = | guesses var_1   errors var_1  |
      | errors var_2    guesses var_2 |
"""

# Plot algorithm results for training set data
from matplotlib.colors import ListedColormap
x_set, y_set = x_train, y_train
x1, x2 = np.meshgrid(np.arange(start = x_set[:,0].min() - 1, stop = x_set[:,0].max() + 1, step = 0.01),
                     np.arange(start = x_set[:,1].min() - 1, stop = x_set[:,1].max() + 1, step = 0.01))
plt.contourf(x1, x2, classifier.predict(np.array([x1.ravel(), x2.ravel()]).T).reshape(x1.shape),
             alpha = 0.75, cmap = ListedColormap(('red', 'green')))
plt.xlim(x1.min(), x1.max())
plt.ylim(x2.min(), x2.max())
for i, j in enumerate(np.unique(y_set)):
    plt.scatter(x_set[y_set == j, 0], x_set[y_set == j, 1],
                c = ListedColormap(('darkred', 'darkgreen'))(i), label = j)
plt.title('Classifier (Some Set)')
plt.xlabel('x_var')
plt.ylabel('y_predictable_var')
plt.legend()
plt.show()

# Reuse classifier.fit() using testing set and plotting algorithm if needed

File: R/Classification/Logistic_Regression.R
# Logistic Regression

# Importar dataset
dataset = read.csv('dataset.csv')
dataset = dataset[,3:5] # Resizing dataset if needed

library(caTools)
set.seed(123)
split = sample.split(dataset$Predictable_var, SplitRatio = 0.75)
training_set = subset(dataset, split == TRUE)
testing_set = subset(dataset, split == FALSE)

# Values Scalating
training_set[,1:2] = scale(training_set[,1:2])
testing_set[,1:2] = scale(testing_set[,1:2])

# Adjust logistic regression model to training set
classifier = glm(formula = Predictable_var ~ .,
                 data = training_set,
                 family = binomial) # Needs to be binomial for yes & no data predictions

# Prediction of the results from the testing set
prob_pred = predict(classifier, type = "response", 
                    newdata = testing_set[,-3])

y_pred = ifelse(prob_pred > 0.5, 1, 0) # 1 & 0 for yes & no

# Create confusion matrix
cm = table(testing_set[,3], y_pred)

# Plotting traning_set. Change the value of var 'set' if needed
library(ElemStatLearn)
set = training_set
x1 = seq(min(set[,1]) - 1, max(set[,1]) + 1, by = 0.01)
x2 = seq(min(set[,2]) - 1, max(set[,2]) + 1, by = 0.01)
grid_set = expand.grid(x1,x2)
colnames(grid_set) = c('Independant_var', 'Dependant_var')
prob_set = predict(classifier, type = 'response', newdata = grid_set)
y_grid = ifelse(prob_set > 0.5, 1, 0)
plot(set[,-3],
     main = 'Clasificación (Testing Set)',
     xlab = 'Independant_var', ylab = 'Dependant_var',
     xlim = range(x1), ylim = range(x2))
contour(x1, x2, matrix(as.numeric(y_grid), length(x1), length(x2)))
points(grid_set, pch = '.', col = ifelse(y_grid == 1,'springgreen3', 'tomato'))
points(set, pch = 21, bg = ifelse(set[,3] == 1, 'green4', 'red3'))

